### PR TITLE
Feature/128 add html handler

### LINKED ISSUE #128: please also add html default handler
https://github.com/biotope/biotope-resource-loader/blob/c7e401dfb3443bfa390d63c2d16c61ab05ec056c/src/defaultOptions.ts#L10

### PR DESCRIPTION
**Pull request by ctmm**
closes #128

### CODE REVIEW

**Comment by timomayer** [comment]
@ctmm can you please rebase this PR after merging #145 #146

**Comment by ctmm** [comment]
@timomayer @SheepFromHeaven, I have updated this branch to reflect the latest changes on master. Ready for review.

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@biotope/resource-loader",
-  "version": "2.1.1",
+  "version": "3.0.0-rc1",
```
Comment (tiagomapmarques) [inline package.json]: This should be done in a separate PR, no?

File: src/ResourceLoader.ts [hunk 2]
```diff
@@ -30,14 +37,33 @@ class ResourceLoader {
...
 		this.prepareQueue(options);
 		loadResources(this.pendingResources);
+	}
+
+	private addMutationObserver(): void {
+		this.mutationObserver = new MutationObserver(this.onMutation.bind(this));
+		this.mutationObserver.observe(this.options.container, this.mutationObserverConfig);
+	}
 
+	private onMutation(mutationRecordArray: MutationRecord[]): void {
```
Comment (tiagomapmarques) [inline src/ResourceLoader.ts]: NIT: IMO should be `handleMutation`

File: src/ResourceLoader.ts [hunk 3]
```diff
@@ -30,14 +37,33 @@ class ResourceLoader {
...
+		mutationRecordArray.forEach(((record: MutationRecord) => {
```
Comment (tiagomapmarques) [inline src/ResourceLoader.ts]: NIT: `record` could be de-constructed

File: src/ResourceLoader.ts [hunk 4]
```diff
@@ -30,14 +37,33 @@ class ResourceLoader {
...
+			record.addedNodes.forEach((node: Node) => {
+				if (node instanceof HTMLElement && node.getAttribute(this.options.resourceListAtrributeSelector)) {
+					this.options = {
```
Comment (tiagomapmarques) [inline src/ResourceLoader.ts]: if `this` refers to the instance of the class, then we should just be adding the `container` attribute, not rewriting the `options` variable altogether (unless there's a valid reason to do it)

File: src/handlers/htmlHandler.ts [hunk 5]
```diff
@@ -0,0 +1,23 @@
+import { HandleOptions } from './../types';
+import { Handler } from '../types';
+
+const isHtml = (options: HandleOptions): boolean => options.resource.path.indexOf('.html') > -1;
```
Comment (tiagomapmarques) [inline src/handlers/htmlHandler.ts]: This will return `true` for this file `file.html.js`, correct? Consider using `slice(-5)` (or `slice(-'.html'.length)`).

File: src/handlers/htmlHandler.ts [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+const onHtmlLoaded = (options: HandleOptions): void => {
+  const elements: HTMLElement[] = [].slice.call(options.resource.elements);
```
Comment (tiagomapmarques) [inline src/handlers/htmlHandler.ts]: Do we need `[].slice.call`? I think `NodeList` has a `forEach` function…

File: src/handlers/htmlHandler.ts [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+  options.response.text()
+    .then(stringOfHtml => {
```
Comment (tiagomapmarques) [inline src/handlers/htmlHandler.ts]: NIT: non-essential new line on the `.then`

File: src/handlers/htmlHandler.ts [hunk 8]
```diff
@@ -0,0 +1,23 @@
+import { HandleOptions } from './../types';
+import { Handler } from '../types';
+
+const isHtml = (options: HandleOptions): boolean => options.resource.path.indexOf('.html') > -1;
+
+const onHtmlLoaded = (options: HandleOptions): void => {
+  const elements: HTMLElement[] = [].slice.call(options.resource.elements);
+
+  options.response.text()
```
Comment (tiagomapmarques) [inline src/handlers/htmlHandler.ts]: Can we return this promise?
(If we need it to be a somewhat "synced" function, we can't do it if we don't return this line)

File: src/defaultOptions.ts [hunk 9]
```diff
@@ -1,6 +1,7 @@
 import cssHandler from './handlers/cssHandler';
 import jsHandler from './handlers/jsHandler';
 import { ResourceLoaderOptions } from './types';
+import htmlHandler from './handlers/htmlHandler';
```
Comment (tiagomapmarques) [inline src/defaultOptions.ts]: NIT: should be together with L1 and L2

File: src/ResourceLoader.ts (hunk 2 above)
Comment (ctmm) [reply src/ResourceLoader.ts]: Updated

File: package.json (hunk 1 above)
Comment (ctmm) [reply package.json]: fixed

File: src/ResourceLoader.ts (hunk 3 above)
Comment (ctmm) [reply src/ResourceLoader.ts]: Sure. Updated

File: src/ResourceLoader.ts (hunk 4 above)
Comment (ctmm) [reply src/ResourceLoader.ts]: Good point. Will fix.

File: src/handlers/htmlHandler.ts (hunk 5 above)
Comment (ctmm) [reply src/handlers/htmlHandler.ts]: I see your point, will update.

I followed this convention because it was already in use for similar .js and .css asset handlers.
We should open a new ticket to update the other handlers.

File: src/handlers/htmlHandler.ts (hunk 6 above)
Comment (ctmm) [reply src/handlers/htmlHandler.ts]: Correct, but the last time I checked Babel did not see fit to poylfill "dom stuff"(their words).
I cast the ```JS NodeList``` into an ```JS Array``` precisely for that reason.

IE11 does not implement the `forEach` method on ```JS NodeList```.

Would prefer to keep as is.

File: src/defaultOptions.ts (hunk 9 above)
Comment (ctmm) [reply src/defaultOptions.ts]: Fixed

File: src/handlers/htmlHandler.ts (hunk 7 above)
Comment (ctmm) [reply src/handlers/htmlHandler.ts]: Ok, will change.

File: src/handlers/htmlHandler.ts (hunk 8 above)
Comment (ctmm) [reply src/handlers/htmlHandler.ts]: I am not seeing a quick way to introduce said functionality because the current logic expects handlers to never return anything.

I haven't yet invested a lot of time on a possible implementation, but it strikes me as something that would take a minimum of half a day effort.

Would this be a deal breaker for merging this PR or is it something we can create an issue for and address later?

**Comment by ctmm** [comment]
@tiagomapmarques I addressed all of your comments, updating the code for small changes. For these small changes I took the liberty of marking the conversation as resolved.

All remarks you made that are open for discussion I left open. 
Kindly ask you to get back at me for the open topics.

Thank you!

File: src/handlers/htmlHandler.ts (hunk 8 above)
Comment (tiagomapmarques) [reply src/handlers/htmlHandler.ts]: Addressing it later is fine :)
